src/orion/core/worker/primary_algo.py

```diff
@@ -79,64 +119,133 @@ def suggest(self, num):
         if transformed_trials is None:
             return None
 
-        for trial in transformed_trials:
-            self._verify_trial(trial, space=self.transformed_space)
-            if trial not in self.transformed_space:
+        trials: list[Trial] = []
+        for transformed_trial in transformed_trials:
+            if transformed_trial not in self.transformed_space:
                 raise ValueError(
-                    f"Trial {trial.id} not contained in space:"
-                    f"\nParams: {trial.params}\n: Space{self.transformed_space}"
+                    f"Trial {transformed_trial.id} not contained in space:\n"
+                    f"Params: {transformed_trial.params}\n"
+                    f"Space: {self.transformed_space}"
                 )
-
-        trials = []
-        for transformed_trial in transformed_trials:
-            trial = self.transformed_space.reverse(transformed_trial)
-            self._verify_trial(trial, space=self.space)
-            trials.append(trial)
-
+            original = self.transformed_space.reverse(transformed_trial)
+            if original in self.registry:
+                # We already have a trial that is equivalent to this one.
+                # Copy over the status and results and observe it.
+                transformed_trial = _copy_status_and_results(
+                    original_trial=original, transformed_trial=transformed_trial
+                )
+                self.algorithm.observe([transformed_trial])
+            else:
+                # We haven't seen this trial before. Register it.
+                trials.append(original)
+
+            # Register the trial in our registry.
+            self.registry.register(original)
+            # Here we register the transformed trial just in case the algorithm hasn't already
+            # registered the trial in its `suggest`.
+            self.algorithm.register(transformed_trial)
```

Comment:
This may result in weird behaviors if the algorithm is not expecting to have trials registered externally. :thinking: I don't have a strong opinion.

Reply:
I guess I'll only call `register` if `has_suggested` is false, would that be better, you think?

Reply:
I'm not sure. I feel like this should be handled by the algorithm itself. If it cannot support observing trials it has not suggested, it should not be fooled by registering the trials externally so that has_suggested() is True.

Reply:
IMO calling `self.algorithm.register` isn't too intrusive (but `self.algorithm.registry.register` would be), since algos can overwrite `register` to only register trials they already have suggested, if they want to.

Anyhoo, what should I be doing here?

Reply:
But they typically register the trial when they suggest it, that's how they know latter on if the suggested the trial. Since we call self.algorithm.observe(trial) I think we do not need to call self.algorithm.register(trial) beforehand. The algo should be responsible of it inside observe.

Reply:
Fixed in https://github.com/Epistimio/orion/pull/833/commits/3c7718784dd1036a8126f87b470acca0220043a1